apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableReadRowsRequestManager.cs

```diff
@@ -0,0 +1,134 @@
+
+        /// <summary>
+        /// Update amount of rows read so far
+        /// </summary>
+        internal void IncrementRowCount(int count) => RowCount += count;
+
+        /// <summary>
+        /// Builds and returns updated subrequest that excludes all rowKeys that have already been found.
+        /// </summary>
+        internal ReadRowsRequest BuildUpdatedRequest()
+        {
+            ReadRowsRequest newReadRowsRequest = new ReadRowsRequest
+            {
+                // Transfer TableName
+                TableNameAsTableName = _originalRequest.TableNameAsTableName,
+                // Transfer RowFilter.
+                Filter = _originalRequest.Filter,
+                // Remove received rowKeys form Rows.
+                Rows = FilterRows()
+            };
+
+            // If the row limit is set, update it.
+            if (_originalRequest.RowsLimit != 0)
+            {
+                long rowsRemaining = _originalRequest.RowsLimit - RowCount;
+                Debug.Assert(rowsRemaining > 0, "The remaining number of rows must be greater than 0.");
+                newReadRowsRequest.RowsLimit = Math.Max(1, rowsRemaining);
+            }
+
+            return newReadRowsRequest;
+        }
+
+        /// <summary>
+        /// Removes received rows from the RowSet.
+        /// </summary>
+        /// <returns>
+        /// New RowSet with not received rows.</returns>
+        private RowSet FilterRows()
+        {
+            RowSet originalRows = _originalRequest.Rows;
+
+            if (LastFoundKey.Value.IsEmpty)
+            {
+                return originalRows;
+            }
+
+            RowSet newRowSet = new RowSet
+            {
+                RowKeys = { originalRows.RowKeys.Where(key => !StartKeyIsAlreadyRead(key)) }
+            };
+
+            foreach (RowRange rowRange in originalRows.RowRanges)
+            {
+                RowRange.EndKeyOneofCase endKeyOneofCase = rowRange.EndKeyCase;
+                if (endKeyOneofCase == RowRange.EndKeyOneofCase.EndKeyClosed &&
+                    EndKeyIsAlreadyRead(rowRange.EndKeyClosed)
+                    || endKeyOneofCase == RowRange.EndKeyOneofCase.EndKeyOpen &&
+                    EndKeyIsAlreadyRead(rowRange.EndKeyOpen))
+                {
+                    continue;
+                }
+
+                RowRange newRange = rowRange;
+                RowRange.StartKeyOneofCase startKeyOneofCase = rowRange.StartKeyCase;
+
+                if (startKeyOneofCase == RowRange.StartKeyOneofCase.StartKeyClosed &&
+                    StartKeyIsAlreadyRead(rowRange.StartKeyClosed)
+                    || (startKeyOneofCase == RowRange.StartKeyOneofCase.StartKeyOpen &&
+                        StartKeyIsAlreadyRead(rowRange.StartKeyOpen))
+                    || startKeyOneofCase == RowRange.StartKeyOneofCase.None)
+                {
+                    newRange.StartKeyOpen = (ByteString) LastFoundKey;
```

Comment:
I just realized this is modifying the original range. You should probably make a clone instead and modify the clone if you get in here.